src/tsolvers/egraph/Explainer.h

```diff
@@ -0,0 +1,101 @@
+//
+// Created by Martin Blicha on 15.09.20.
+//
+
+#ifndef OPENSMT_EXPLAINER_H
+#define OPENSMT_EXPLAINER_H
+
+#include "EnodeStore.h"
+#include "UFInterpolator.h"
+#include <memory>
+
+class Explainer {
+protected:
+    //
+    // Fast duplicates checking. Cannot be nested !
+    //
+    struct DupChecker;
+    class DuplicateCheckerData {
+        Map<PTRef,int,PTRefHash>    duplicates;                       // Fast duplicate checking
+        int                         dup_count = 0;                    // Current dup token
+        bool                        free = true;
+        friend                      struct DupChecker;
+    };
+
+    DuplicateCheckerData duplicateCheckerData;
+
+    struct DupChecker {
+        DuplicateCheckerData &dc;
+        DupChecker(DuplicateCheckerData &dc) : dc(dc) {
+            if (!dc.free) {
+                throw OsmtInternalException(); // "Attempt to re-use DuplicateChecker without releasing"
+            }
+            dc.free = false;
+            if (dc.dup_count < std::numeric_limits<int>::max()) {
+                ++dc.dup_count;
+            } else {
+                dc.duplicates.clear();
+            }
+        }
+        inline void storeDup(PTRef e) { assert(!dc.free); if (dc.duplicates.has(e)) dc.duplicates[e] = dc.dup_count; else dc.duplicates.insert(e, dc.dup_count); }
+        inline bool isDup   (PTRef e) { assert(!dc.free); return !dc.duplicates.has(e) ? false : dc.duplicates[e] == dc.dup_count; }
+        ~DupChecker() {
+            dc.free = true;
+        }
+    };
+
+    DuplicateCheckerData dcd;
```

Comment:
There seem to be 2 data members of type DuplicateCheckerData: `dcd` and `duplicateCheckerData`

Reply:
Removed the redundant one